lbrynet/core/HTTPBlobDownloader.py

```diff
@@ -0,0 +1,79 @@
+from random import choice
+import logging
+
+from twisted.internet import defer
+import treq
+from twisted.internet.task import LoopingCall
+
+log = logging.getLogger(__name__)
+
+
+class HTTPBlobDownloader(object):
+    def __init__(self, blob_manager, blob_hashes=None, servers=None, client=None):
+        self.blob_manager = blob_manager
+        self.servers = servers or []
+        self.client = client or treq
+        self.blob_hashes = blob_hashes or []
+        self.looping_call = LoopingCall(self._download_next_blob_hash_for_file)
+        self.failures = 0
+        self.max_failures = 3
+        self.interval = 1
+
+    @property
+    def running(self):
+        return self.looping_call.running
+
+    def start(self):
+        if not self.running and self.blob_hashes and self.servers:
+            return self.looping_call.start(self.interval, now=True)
+        defer.succeed(None)
+
+    def stop(self):
+        if self.running:
+            self.blob_hashes = []
+            return self.looping_call.stop()
+
+    @defer.inlineCallbacks
+    def _download_next_blob_hash_for_file(self):
+        for blob_hash in self.blob_hashes:
+            blob = yield self.blob_manager.get_blob(blob_hash)
+            if not blob.verified:
+                self.download_blob(blob)
+                return
+        self.stop()
+
+    @defer.inlineCallbacks
+    def download_blob(self, blob):
+        try:
+            blob_hash = yield self._download_blob(blob)
+            if blob_hash:
+                log.debug('Mirror completed download for %s', blob_hash)
+            self.failures = 0
+        except Exception as exception:
```

Comment:
all exceptions? shouldnt we be more specific?

Reply:
We don't have a way to know what are all of the possible exceptions that could be thrown. I think the goal here is to retry downloading no matter what and log the specific exception.

Reply:
`exception` variable still holds the actual specific exception, the error message will include the specific exception in the output. This is basically a catch all to make sure we keep jugging along no matter what might have failed.

Reply:
Yes, in general, but I believe not in this case. This loop doesn't raise to callers, so if the exception isn't handled and logged here it won't be anywhere and could cause cryptic "unhandled error" messages. IMO, but if you disagree we can be specific.